lib/compat.php

```diff
@@ -304,6 +304,261 @@ function gutenberg_warn_classic_about_blocks() {
+ * attributes for front-end display.
+ *
+ * @since 4.5.0
+ *
+ * @param string $html  Original HTML.
+ * @param array  $block Parsed block.
+ * @return string Filtered Image block HTML.
+ */
+function gutenberg_render_block_core_image( $html, $block ) {
+	// Old post or... something's wrong.
+	if ( empty( $html ) || empty( $block['attrs'] ) ) {
+		return $html;
+	}
+
+	$defaults = array(
+		'url'   => '',
+		'alt'   => '',
+		'id'    => 0,
+		'align' => '',
+	);
+
+	$block_attributes = wp_parse_args( $block['attrs'], $defaults );
+
+	if ( empty( $block_attributes['url'] ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	if ( ! empty( $block_attributes['id'] ) ) {
+		$attachment_id = (int) $block_attributes['id'];
+		$image_meta    = wp_get_attachment_metadata( $attachment_id );
+	} else {
+		$attachment_id = 0;
+		$image_meta    = null;
+	}
+
+	$image_dimensions = gutenberg_get_image_width_height( $block_attributes, $image_meta );
+	$image_src        = '';
+	$srcset           = '';
+	$sizes            = '';
+
+	if ( empty( $image_dimensions ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	$image_src = $block_attributes['url'];
+
+	$image_attributes = array(
+		'src'    => $image_src,
+		'alt'    => empty( $block_attributes['alt'] ) ? '' : $block_attributes['alt'],
+		'width'  => $image_dimensions[0],
+		'height' => $image_dimensions[1],
+	);
+
+	if ( $image_meta ) {
+		// TODO: pass `$block_attributes` to the filter.
+		$srcset = wp_calculate_image_srcset( $image_dimensions, $image_src, $image_meta, $attachment_id );
+
+		if ( ! empty( $srcset ) ) {
+			// TODO: pass `$block_attributes` to the filter. This will let themes generate better `sizes` attribute.
+			$sizes = wp_calculate_image_sizes( $image_dimensions, $image_src, $image_meta, $attachment_id );
+		}
+
+		if ( $srcset && $sizes ) {
+			$image_attributes['srcset'] = $srcset;
+			$image_attributes['sizes']  = $sizes;
+		}
+	}
+
+	/**
+	 * Filters the image tag attributes when rendering the core image block.
+	 *
+	 * @since 4.5.0
+	 *
+	 * @param array  $image_attributes The (recalculated) image attributes.
+	 *                                 Note: expects valid HTML 5.0 attribute names.
+	 * @param array  $block_attributes The image block attributes.
+	 * @param string $html             The image block HTML coming from the
+	 *                                 editor. The img tag will be replaced.
+	 */
+	$image_attributes = apply_filters( 'render_block_core_image_tag_attributes', $image_attributes, $block_attributes, $html );
+
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$is_invalid_attribute = preg_match( '/[^a-z0-9-]+/', '', strtolower( $name ) );
```

Comment:
(copying here because it was hidden through a code update)

> The only exception (well, more of a limitation) may be for data-* attribute names

If we want to conform to the HTML spec we can do just that - it's not hard

> Attributes have a name and a value. Attribute names must consist of one or more characters other than the space characters, U+0000 NULL, U+0022 QUOTATION MARK ("), U+0027 APOSTROPHE ('), U+003E GREATER-THAN SIGN (>), U+002F SOLIDUS (/), and U+003D EQUALS SIGN (=) characters, the control characters, and any characters that are not defined by Unicode. In the HTML syntax, attribute names, even those for foreign elements, may be written with any mix of lower- and uppercase letters that are an ASCII case-insensitive match for the attribute’s name.
> https://www.w3.org/TR/html52/syntax.html#attribute-names

```js
/[^\u0000-\u0020\u007f-\u009f"'>/=]+/
```

that should define them. that leaves open some valid names that WordPress will sanitize, but they are valid.

Reply:
The intention there was to check for valid HTML 5.0 attribute names only, not all possible attribute names. In HTML 5.0 the only attribute name that may contain any of the valid chars is the `data-*` (in the variable part).

However I think it's better to keep that variable part limited to only `a-z0-9-` like it is at other places in core.

As discussed previously, it would be good to make that into another "filtering" function in core. Perhaps then we can decide whether to support all valid chars or to keep the restriction.